pallets/connectors/src/lib.rs

```diff
@@ -480,143 +533,359 @@ pub mod pallet {
+		) -> Result<CurrencyIdOf<T>, DispatchError> {
+			let currency = CurrencyIdOf::<T>::try_from(index)?;
+			ensure!(
+				T::AssetRegistry::metadata(&currency).is_some(),
+				Error::<T>::AssetNotFound
+			);
+
+			Ok(currency)
+		}
+
+		/// Checks whether the given currency is transferable via Connectors and
+		/// whether its metadata contains a location which can be
+		/// converted to [ConnectorsWrappedToken].
+		///
+		/// Requires the currency to be registered in the `AssetRegistry`.
+		pub fn try_get_wrapped_token(
+			currency_id: &CurrencyIdOf<T>,
+		) -> Result<ConnectorsWrappedToken, DispatchError> {
+			let meta = T::AssetRegistry::metadata(currency_id).ok_or(Error::<T>::AssetNotFound)?;
+			ensure!(
+				meta.additional.transferability.includes_connectors(),
+				Error::<T>::AssetNotConnectorsTransferable
+			);
+
+			match meta.location {
+				Some(VersionedMultiLocation::V3(MultiLocation {
+					parents: 0,
+					interior:
+						X3(
+							PalletInstance(pallet_instance),
+							GlobalConsensus(NetworkId::Ethereum { chain_id }),
+							AccountKey20 {
+								network: None,
+								key: address,
+							},
+						),
+				})) if Some(pallet_instance.into())
+					== <T as frame_system::Config>::PalletInfo::index::<Pallet<T>>() =>
+				{
+					Ok(ConnectorsWrappedToken::EVM { chain_id, address })
+				}
+				_ => Err(Error::<T>::AssetNotConnectorsWrappedToken.into()),
+			}
+		}
+
+		/// Ensures that the given pool and tranche exists and returns the
+		/// corresponding investment id.
+		pub fn derive_invest_id(
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+		) -> Result<<T as pallet::Config>::TrancheCurrency, DispatchError> {
+			ensure!(
+				T::PoolInspect::pool_exists(pool_id),
+				Error::<T>::PoolNotFound
+			);
+			ensure!(
+				T::PoolInspect::tranche_exists(pool_id, tranche_id),
+				Error::<T>::TrancheNotFound
+			);
+
+			Ok(TrancheCurrency::generate(pool_id, tranche_id))
+		}
+
+		/// Ensures that the payment currency of the given investment id matches
+		/// the derived currency and returns the latter.
+		pub fn try_get_payment_currency(
+			invest_id: <T as pallet::Config>::TrancheCurrency,
+			currency_index: GeneralCurrencyIndexOf<T>,
+		) -> Result<CurrencyIdOf<T>, DispatchError> {
+			// retrieve currency id from general index
+			let currency = Self::try_get_currency_id(currency_index)?;
+
+			ensure!(
+				T::ForeignInvestment::accepted_payment_currency(invest_id, currency),
+				Error::<T>::InvalidInvestCurrency
+			);
+
+			Ok(currency)
+		}
+	}
+
+	impl<T: Config> InboundQueue for Pallet<T>
+	where
+		<T as frame_system::Config>::AccountId: From<[u8; 32]>,
+	{
+		type Message = MessageOf<T>;
+		type Sender = DomainAddress;
+
+		#[transactional]
+		fn process(sender: DomainAddress, msg: MessageOf<T>) -> DispatchResult {
+			ensure!(
+				KnownConnectors::<T>::contains_key::<&T::AccountId>(&sender.address().into()),
+				Error::<T>::InvalidIncomingMessageOrigin
+			);
```

Comment:
Not needed. All validation of message submitters happen in the Gatway. Must not be addressed here. Can also be addressed by the gatway PR.